diff --git a/megatron/core/models/common/embeddings/rope_utils.py b/megatron/core/models/common/embeddings/rope_utils.py
index e39540eb1..70783770d 100644
--- a/megatron/core/models/common/embeddings/rope_utils.py
+++ b/megatron/core/models/common/embeddings/rope_utils.py
@@ -42,9 +42,30 @@ __all__ = [
     'fused_apply_rotary_pos_emb',
     'fused_apply_rotary_pos_emb_thd',
     'get_pos_emb_on_this_cp_rank',
+    'apply_rotary_pos_emb_bshd_in_complex',
 ]
 
 
+def apply_rotary_pos_emb_bshd_in_complex(
+ 	     t: Tensor,
+ 	     freqs: Tensor,
+ 	     rotary_interleaved: bool = False
+ 	 ) -> Tensor:
+ 	if rotary_interleaved:
+ 	    s, b, n, d = t.shape
+ 	    t = t.view(s, b, n, 2, d // 2).transpose(4, 3)
+ 	freqs, _ = freqs.chunk(2, dim=-1)
+
+ 	freqs_cis = torch.polar(torch.ones_like(freqs), freqs)
+
+ 	x = torch.view_as_complex(t.float().view(*t.shape[:-1], -1, 2)).squeeze(-1)
+ 	y = torch.view_as_real(x * freqs_cis).flatten(3)
+ 	if rotary_interleaved:
+ 	    y = torch.cat([y[..., 0::2], y[..., 1::2]], dim=-1)
+
+ 	return y.to(t.dtype)
+
+
 def get_pos_emb_on_this_cp_rank(
     pos_emb: Tensor, seq_dim: int, cp_group: torch.distributed.ProcessGroup
 ) -> Tensor:
diff --git a/megatron/core/transformer/experimental_attention_variant/dsa.py b/megatron/core/transformer/experimental_attention_variant/dsa.py
index 353b31e9b..fc56b609a 100644
--- a/megatron/core/transformer/experimental_attention_variant/dsa.py
+++ b/megatron/core/transformer/experimental_attention_variant/dsa.py
@@ -13,6 +13,7 @@ from megatron.core.models.common.embeddings import (
     YarnRotaryEmbedding,
     apply_rotary_pos_emb,
 )
+from megatron.core.models.common.embeddings.rope_utils import apply_rotary_pos_emb_bshd_in_complex
 from megatron.core.packed_seq_params import PackedSeqParams
 from megatron.core.process_groups_config import ProcessGroupCollection
 from megatron.core.tensor_parallel.mappings import gather_from_sequence_parallel_region
@@ -371,9 +372,9 @@ class DSAIndexer(MegatronModule):
             else self.config.hidden_size
         )
 
-        self.index_n_heads = self.config.dsa_indexer_n_heads
-        self.index_head_dim = self.config.dsa_indexer_head_dim
-        self.index_topk = self.config.dsa_indexer_topk
+        self.index_n_heads = 64
+        self.index_head_dim = 128
+        self.index_topk = 2048
 
         self.softmax_scale: float = self.index_head_dim**-0.5
 
@@ -456,19 +457,15 @@ class DSAIndexer(MegatronModule):
         """Apply RoPE to the input tensor."""
         # x_nope [seqlen, batch, *, index_head_dim - qk_pos_emb_head_dim]
         # x_pe   [seqlen, batch, *, qk_pos_emb_head_dim]
-        x_nope, x_pe = torch.split(
-            x, [self.index_head_dim - self.qk_pos_emb_head_dim, self.qk_pos_emb_head_dim], dim=-1
+        x_pe, x_nope = torch.split(
+            x, [self.index_head_dim - self.qk_pos_emb_head_dim,self.qk_pos_emb_head_dim], dim=-1
         )
-        x_pe = apply_rotary_pos_emb(
+        x_pe = apply_rotary_pos_emb_bshd_in_complex(
             x_pe,
             rotary_pos_emb,
-            config=self.config,
-            cu_seqlens=None,
-            mscale=mscale,
-            cp_group=self.pg_collection.cp,
-        )
+            rotary_interleaved=True)
         # [seqlen, batch, *, index_head_dim]
-        x = torch.cat([x_nope, x_pe], dim=-1)
+        x = torch.cat([x_pe, x_nope], dim=-1)
         return x
 
     def _compute_index_scores(
