diff --git a/recipe/dapo/main_dapo.py b/recipe/dapo/main_dapo.py
index ffd1e438..87f592bd 100644
--- a/recipe/dapo/main_dapo.py
+++ b/recipe/dapo/main_dapo.py
@@ -142,7 +142,9 @@ class TaskRunner:
             max_resp_len=config.data.max_response_length,
             overlong_buffer_cfg=config.reward_model.overlong_buffer,
         )
-
+        if config.actor_rollout_ref.rollout.partial_rollout_mode == "async" and config.actor_rollout_ref.rollout.partial_rollout_max_split > 0:
+            from verl.trainer.ppo.ray_trainer import AggregatorActor
+            role_worker_mapping[Role.Aggregator] = ray.remote(AggregatorActor)
         # Note that we always use function-based RM for validation
         val_reward_fn = load_reward_manager(
             config,
