diff --git a/verl/workers/megatron_workers.py b/verl/workers/megatron_workers.py
index a89cff18..1c179d0a 100644
--- a/verl/workers/megatron_workers.py
+++ b/verl/workers/megatron_workers.py
@@ -162,8 +162,10 @@ class MegatronWorker(Worker):
             assert use_mbridge, "fp16 mode requires use_mbridge to be True"
         if use_mbridge:
             from verl.models.mcore.mbridge import AutoBridge
+            from verl.models.mcore.patch_v012 import apply_patch
+            apply_patch()
 
-            bridge = AutoBridge.from_config(hf_config, dtype=dtype)
+            bridge = AutoBridge.from_config(hf_config)
             bridge.set_extra_args(**override_transformer_config)
             tf_config = bridge.config
             tf_config.fp16 = fp16
