diff --git a/mindspeed/core/transformer/experimental_attention_variant/dsa_fused.py b/mindspeed/core/transformer/experimental_attention_variant/dsa_fused.py
index af073b81..31e73504 100644
--- a/mindspeed/core/transformer/experimental_attention_variant/dsa_fused.py
+++ b/mindspeed/core/transformer/experimental_attention_variant/dsa_fused.py
@@ -13,7 +13,7 @@ try:
 except ImportError:
     HAVE_EINOPS = False
 
-from megatron.training import get_args
+from mindspeed.args_utils import get_full_args as get_args
 from megatron.core.transformer.experimental_attention_variant.dsa import (
     DSAIndexerLossLoggingHelper,
     unfused_dsa_fn,
diff --git a/mindspeed/features_manager/transformer/experimental_attention_variant/dsa.py b/mindspeed/features_manager/transformer/experimental_attention_variant/dsa.py
index 22e55917..c8431581 100644
--- a/mindspeed/features_manager/transformer/experimental_attention_variant/dsa.py
+++ b/mindspeed/features_manager/transformer/experimental_attention_variant/dsa.py
@@ -46,7 +46,8 @@ class DeepSeekSparseAttention(MindSpeedFeature):
             pm.register_patch('megatron.core.models.gpt.experimental_attention_variant_module_specs.get_dsa_module_spec_for_backend', get_dsa_module_spec_for_backend)
 
         from mindspeed.core.transformer.experimental_attention_variant.dsa_fused import forward_with_scores, fused_dsa_attn_forward
-        pm.register_patch('megatron.core.transformer.experimental_attention_variant.dsa.DSAIndexer.forward_with_scores', forward_with_scores)
+        from megatron.core.transformer.experimental_attention_variant.dsa import DSAIndexer
+        DSAIndexer.forward_with_scores = forward_with_scores
         pm.register_patch('megatron.core.transformer.experimental_attention_variant.dsa.DSAttention.forward', fused_dsa_attn_forward)
 
         if int(getattr(args, 'context_parallel_size', 1)) > 1 and getattr(args, 'context_parallel_algo', 'megatron_cp_algo') == 'kvallgather_cp_algo':
