diff --git a/vllm_ascend/ops/fused_moe/moe_comm_method.py b/vllm_ascend/ops/fused_moe/moe_comm_method.py
index c48ce1a4..ecbc8454 100644
--- a/vllm_ascend/ops/fused_moe/moe_comm_method.py
+++ b/vllm_ascend/ops/fused_moe/moe_comm_method.py
@@ -88,6 +88,8 @@ class MoECommMethod(ABC):
             topk_weights: torch.Tensor,
             topk_ids: torch.Tensor,
             activation: str = "silu",
+            w1_bias: Optional[torch.Tensor] = None,
+            w2_bias: Optional[torch.Tensor] = None,
             apply_router_weight_on_input: bool = False,
             use_int8_w8a8: bool = False,
             use_int4_w4a8: bool = False,
@@ -142,6 +144,9 @@ class MoECommMethod(ABC):
                                        w1_scale=w1_scale,
                                        w2=w2,
                                        w2_scale=w2_scale,
+                                       w1_bias=w1_bias,
+                                       w2_bias=w2_bias,
+                                       activation=activation,
                                        group_list=expert_tokens,
                                        dynamic_scale=dynamic_scale,
                                        group_list_type=group_list_type,
