diff --git a/vllm_ascend/utils.py b/vllm_ascend/utils.py
index a539c5a3..9f69d423 100644
--- a/vllm_ascend/utils.py
+++ b/vllm_ascend/utils.py
@@ -63,6 +63,15 @@ _HAS_LAYER_IDX = None
 _SUBSCRIBED_COMPUTE_STREAMS = set()
 _GRAPH_PRINT_STREAM = None
 _GRAPH_PRINT_STREAM_LOCK = Lock()
+_USE_SINK_ATTENTION = False
+
+def set_use_sink_attention(value: bool):
+    global _USE_SINK_ATTENTION
+    _USE_SINK_ATTENTION = value
+
+def get_use_sink_attention() -> bool:
+    global _USE_SINK_ATTENTION
+    return _USE_SINK_ATTENTION
 
 
 def _print_callback_on_stream(*args):
